vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBox.java

```diff
@@ -618,8 +967,14 @@ protected void onDetach(DetachEvent detachEvent) {
     }
 
     private void refreshAllData(boolean forceServerSideFiltering) {
-        setClientSideFilter(!forceServerSideFiltering && getDataProvider()
-                .size(new Query<>()) <= getPageSizeDouble());
+        setClientSideFilter(!forceServerSideFiltering
+                && dataCommunicator.getItemCount() <= getPageSizeDouble()
+                // We do want to perform server side filtering (send client
+                // filter to the server) if there is at least one item count
+                // change listener added, even in case of in-memory data
+                // provider and even if the item count < page size.
+                // This allows us be always notified about size change.
+                && itemCountListeners.get() == 0);
```

Comment:
As commented about this, it feels odd behavior change, that could be avoided with just delegating the client side filtering result count to server side instead

Reply:
This code has been removed and item count notification now goes from client to server side in client filter mode.

Reply:
> item count notification now goes from client to server side in client filter mode.

It has been removed too. Client filter change should not trigger item count change event on the server. But this will be a separate ticket.